src/backend/buffer.rs

```diff
@@ -123,30 +123,39 @@ where
         unsafe { str::from_utf8_unchecked(str_bytes) }
     }
 
-    /// Pushes the given value onto the buffer with `var7` encoding.
+    /// Pushes the given length value onto the buffer with `var7` encoding.
+    /// Ensures there's enough capacity for the `var7` encoded length and
+    /// the following string bytes ahead of pushing.
     ///
     /// Returns the amount of `var7` encoded bytes.
     #[inline]
-    fn encode_var_usize(&mut self, value: usize) -> usize {
-        encode_var_usize(&mut self.buffer, value)
+    fn try_encode_var_length(&mut self, length: usize) -> Result<()> {
+        let add_len = length + calculate_var7_size(length);
```

Comment:
I would rename it to something like `len_encode_var_usize` which is very similar to `encode_var_usize` to draw the parallels to the reader of the code.

Reply:
Though I have to admit that it is a bit unfortunate that this encoding step is now 2 phased.

Reply:
`encode_var_usize` pushes each byte separately, calling `try_reserve` every time a byte is encoded would take down the performance even more. However with `push_within_capacity` stabilized that could be implemented differently.

Alternative solution is to encode length into an array on the stack and then call `extend_from_slice` on `self.buffer`. This however can't be done efficiently without a bit of unsafe code (`MaybeUninit`) or would require using ArrayVec or a similar crate.